### PR TITLE
MAINT: remove exec_command from gnu.py

### PR DESCRIPTION
**Pull request by tylerjereddy**
Related to #11824

### CODE REVIEW

File: numpy/distutils/fcompiler/gnu.py
Comment (eric-wieser) [inline numpy/distutils/fcompiler/gnu.py]: Nit: Move all but the first line to an `else`

**Comment by eric-wieser** [comment]
What's going on with the weird `bytes --(preferred)--> unicode --(ascii)--> bytes` conversion here? Was that contained in `exec_command`? Sounds like it will break completely on unicode paths.

**Comment by tylerjereddy** [comment]
>  Was that contained in exec_command?

Yeah, I made a comment about this in one of the PRs on this topic--sometimes my replacement code broke until I put that back in from the part of `exec_command` that happens after the `Popen()`--I may not actually need **all** of it, but I guess I'm being conservative given how awkward that stuff is to test, and to test for platform portability.

File: numpy/distutils/fcompiler/gnu.py
Comment (eric-wieser) [reply numpy/distutils/fcompiler/gnu.py]: The `str` here is a no op on both python versions

**Comment by tylerjereddy** [comment]
Revised & added comment for historical string processing block

File: numpy/distutils/fcompiler/gnu.py
Comment (eric-wieser) [inline numpy/distutils/fcompiler/gnu.py]: Isn't the same historical processing used here? Perhaps you should add a `filepath_from_subprocess_output` function that you use in all these places, and you can stick the comments in. Something like
```
"""
Convert `bytes` in the encoding used by a subprocess into a filesystem-appropriate `str`.

Inherited from `exec_command`, and possibly incorrect.
"""
```

**Comment by tylerjereddy** [comment]
Revised with suggested function abstraction; the `locale` import is probably spurious now; may be safer to use `filepath_from_subprocess_output` in all locations for now, even if the CI is green with omissions at select usages.

File: numpy/distutils/fcompiler/gnu.py
Comment (eric-wieser) [reply numpy/distutils/fcompiler/gnu.py]: I don't see any reason why these two cases would be different - both are asking the same executable for a filename

**Comment by tylerjereddy** [comment]
Revised to use the abstracted function for all locations in this PR

File: numpy/distutils/exec_command.py
Comment (eric-wieser) [inline numpy/distutils/exec_command.py]: Can you preserve all the comments from the original in `_exec_command` here?

File: numpy/distutils/fcompiler/gnu.py
Comment (eric-wieser) [inline numpy/distutils/fcompiler/gnu.py]: There's a long comment in `exec_command` about passing `universal_newlines=False` here - are you sure it was incorrect?

**Comment by eric-wieser** [comment]
@tylerjereddy: Would be good to finish this up so that the others can go in too.

File: numpy/distutils/fcompiler/gnu.py
Comment (tylerjereddy) [reply numpy/distutils/fcompiler/gnu.py]: I think it is probably correct but `universal_newlines=False` is the default for `subprocess.check_output`, so perhaps less convincing that we still need a paragraph explaining why the default value is used

**Comment by tylerjereddy** [comment]
I restored the historical comments in `filepath_from_subprocess_output` as requested

File: numpy/distutils/fcompiler/gnu.py
Comment (eric-wieser) [reply numpy/distutils/fcompiler/gnu.py]: Ah, the default is `False`, so this is fine

**Review by eric-wieser — APPROVED** [review]
Looks good now, thanks. Waiting on CI, but I doubt it exercises any of this anyway. Feel free to merge this and rebase the others once this goes green.

**Comment by tylerjereddy** [comment]
The Travis matrix looks all green, but it doesn't seem to want to tell github about it.

**Comment by charris** [comment]
Might add a release note, maybe under "Improvements", to the effect that builds should now be independent of the shell. I'll put this in and you can tack that onto one of the other PRs.

**Comment by charris** [comment]
Thanks Tyler.

**Pull request merged**
